### DISCUSSION TITLE
v41.0

### DISCUSSION THREAD

**Discussion by westnordost** [streetcomplete/StreetComplete#3819] [Release Discussions]
![Screenshot_20220225-153818_StreetComplete](https://user-images.githubusercontent.com/4661658/155899281-0ab790ac-978d-4107-b6b9-460b8741c29c.jpg)

  <h3>Measure with the camera (AR)</h3>
  <p>Measuring with AR is only supported <a href="https://developers.google.com/ar/devices#google_play_devices">on certain devices</a> and you need to have the Google Play Services on your phone. If AR is not available, the new quests are disabled by default because you'd need to use a tape measure instead. (#3709, #879)</p>
  <h3>New Quests</h3>
  <ul>
  <li><i>What’s the width of this cycleway here?</i> (#698)</li>
  <li><i>What’s the width of this road here?</i></li>
  <li><i>What is the height limit here?</i></li>
  </ul>
  <h3>Quest enhancements</h3>
  <ul>
  <li>Add "handlebar holders" as bicycle parking type option (#3061, #3724), by @sams96</li>
  <li>Tapping on road in street name sometimes did not work (#3797)</li>
  <li>Other small improvements (#3740, #3755, #3793, #3783, #3814, #3811), thanks @matkoniecz, @tiptoptom, @peternewman, @arrival-spring</li>
  </ul>

<hr /><em>This discussion was created from the release <a href='https://github.com/streetcomplete/StreetComplete/releases/tag/v41.0'>v41.0</a>.</em>

**Comment by mnalis** [comment]
has something changed about camera permissions recently? 
Why I try to build Dev version `.apk`, I can no longer leave photos in notes - I get `Unable to create file for photo` message:

<Details>

```
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment: Unable to create photo
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment: java.lang.SecurityException: Permission Denial: starting Intent { act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.huawei.camera/.ThirdCamera clip={text/uri-list=1 U:content://de.westnordost.streetcomplete.mn.debug.fileprovider/images/photo_1646019765120.jpg} (has extras) } from ProcessRecord{b40b5d2 19653:de.westnordost.streetcomplete.mn.debug/u0a407} (pid=19653, uid=10407) with revoked permission android.permission.CAMERA
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at android.os.Parcel.createException(Parcel.java:2071)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at android.os.Parcel.readException(Parcel.java:2039)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at android.os.Parcel.readException(Parcel.java:1987)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at android.app.IActivityTaskManager$Stub$Proxy.startActivity(IActivityTaskManager.java:3888)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at android.app.Instrumentation.execStartActivity(Instrumentation.java:1738)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at android.app.Activity.startActivityForResult(Activity.java:5368)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:588)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at androidx.core.app.ActivityCompat.startActivityForResult(ActivityCompat.java:239)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at androidx.activity.ComponentActivity$2.onLaunch(ComponentActivity.java:211)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at androidx.activity.result.ActivityResultRegistry$2.launch(ActivityResultRegistry.java:167)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at androidx.fragment.app.Fragment$9.launch(Fragment.java:3507)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at androidx.activity.result.ActivityResultLauncher.launch(ActivityResultLauncher.java:47)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at de.westnordost.streetcomplete.util.ActivityForResultLauncher.invoke(ActivityForResultLauncher.kt:20)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at de.westnordost.streetcomplete.quests.note_discussion.AttachPhotoFragment.takePhoto(AttachPhotoFragment.kt:74)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at de.westnordost.streetcomplete.quests.note_discussion.AttachPhotoFragment.access$takePhoto(AttachPhotoFragment.kt:31)
02-28 04:42:45.126 19653 19653 E AttachPhotoFragment:   at de.westnordost.streetcomplete.quests.note_discussion.AttachPhotoFragment$onViewCreated$1$1.invokeSuspend(AttachPhotoFragment.kt:52)

```
</details>

I've never had that problem with `39.x` and previous versions and camera worked normally there. I also think (but only 95% sure) that most of `40.x`series also worked fine.

Now I have to go to android system permissions where `Camera` permission is shown as `Denied` and I have to allow it manually, then it works (until next time that I uninstall / reinstall devel version when it breaks again)

**Reply by westnordost** [reply]
Hm, great. So apparently, when an Android app doesn't request the camera permission in its manifest, it may request the camera app to take a picture freely. But when an Android app **does** request the camera permission and the user denied that camera permission, it may not request the camera app to take a picture anymore.

**Reply by westnordost** [reply]
Latest commit should fix this (ask for camera permission).

**Reply by mnalis** [reply]
Yes, it fixes this for me

**Comment by peternewman** [comment]
I can't even install v41.0 :sob: .

If I go to:
https://play.google.com/store/apps/details?id=de.westnordost.streetcomplete

It says:
Requires Android
5.0 and up

But drilling into the detail of choose a device says "This item is not compatible with your device".

I'm currently on v40.2, I never saw the v41.0-beta1 (did it actually make it to GitHub), but I was able to install v40.0-beta1 with the AR stuff you had in a PR before. My phone is actually on Android 9.

I assume having the Dev version installed too won't be breaking it? I can't remember if I've upgraded since that, but presumably given I've got v40.2 now?

**Reply by westnordost** [reply]
Hm. And when you download the app from GitHub? ArCore usage, OpenGL ES 3.1 usage, camera usage and even GPS usage have been marked as "optional" in the manifest.

**Reply by peternewman** [reply]
I haven't tried yet, but when I tried here ( https://github.com/streetcomplete/StreetComplete/pull/3709#issuecomment-1028541887 ) it worked, and not much seems to have changed since then...

**Reply by peternewman** [reply]
I just get a "This App Was Not Installed" message (as opposed to my usual you don't have enough space one).

**Reply by westnordost** [reply]
Is it possible for you to get the reason somehow for why it was not installed? Maybe something in the system log?

**Reply by westnordost** [reply]
Note that if you use the version from F-Droid, you must first deinstall that version to be able to install a version from Google Play or GitHub.

**Reply by matkoniecz** [reply]
See also https://github.com/streetcomplete/StreetComplete/issues/3820

**Reply by matkoniecz** [reply]
I get 

> This app is no longer compatible with your device. Contact the developers for more info

in play store for my device.

**Reply by peternewman** [reply]
I'm not sure what's going on here, I had to restart as I couldn't get adb to connect, at which point the GitHub APK has installed fine, but Play Store still says it's not compatible...

So v41.0 is now on, and AR measuring works, but I seem to be outside of the app store...

**Reply by peternewman** [reply]
> See also #3820

Glad it's not just me, and I am AR capable FWIW!

**Comment by vizsim** [comment]
Hi,
i just added the width of a [highway](https://www.openstreetmap.org/way/155605328) via StreetComplete, then realized, that the tag ["width:carriageway"](https://wiki.openstreetmap.org/wiki/Key:width:carriageway) has already been assigned.

Which key is preferable? Is it possible to add "width:carriageway" in SC so duplicates of width can be avoided?

**Reply by FloEdelmann** [reply]
If `width:carriageway` is set, the `AddRoadWidth` quest will update it (while also setting `width`):

https://github.com/streetcomplete/StreetComplete/blob/79557f0fef48f4e348b9602b25299eeb6b871712/app/src/main/java/de/westnordost/streetcomplete/quests/width/AddRoadWidth.kt#L55-L56

**Reply by kmpoppe** [reply]
The `elementFilter` for `AddRoadWidth` does not exclude `width:carriageway` - which I think is reasonable with only 1.8k uses of that tag - and as @FloEdelmann states, it's set again which I think is also very smart because the quest specifically asks you to measure the curb to curb distance which is the whole sence for  `width:carriageway`.

**Reply by vizsim** [reply]
Alright thanks, I guess I am just a bit confused how these two tags ( `width:carriageway` and `width`) actually differ, especially if parking is not really clearly a curb.
Btw the refering street looks like this:
https://www.mapillary.com/app/?lat=52.479553699972&lng=13.4207476&z=19.9&pKey=409097403889651&focus=photo&x=0.45489912723893206&y=0.49835855243522853&zoom=0
